### PR TITLE
Single Initialization: Removes the current progressive initialization in favor of a single underlying call to init_step.

### PR DESCRIPTION
**Pull request by cgarciae**
Currently we implement a progressive initialization scheme where we keep track of the current `init_stage` which could be `{none, pred, test, train}`, these are ordered e.g. if current `init_stage` is `test` and `predict` is called no initialization occurs, but if `fit` is called initialization is called and `init_stage` is set to `train` which is the final state. 

The previous scheme had the benefit that you could do things like:

```python
model = Model(...)

# call predict immediately, net_params and net_states are initialized
y_pred = model.predict(x)

# now call fit, metrics_states and optimizer_states are initialized
model.fit(x, y)
```
However it has the following downsides:
* Initialization is harder to reason about since there are potentially multiple passes.
* Order of calls to `predict`, `evaluate` or `fit` could potentially give rise to different initial parameters.
* Forced additional compilations.

## Proposal
Replace the progressive initialization scheme with a single initialization scheme, this means refactoring `init_stage: Mode` to `initialized: bool` and making the default implementation of `init_step` call `train_step` directly to completely initialize the `Model`.

#### Implications
* You can still call `fit`, and `evaluate` directly because they have all the information needed to initialize the full model (`x` and `y_true`).
* Now you cannot call `predict` directly, you must call `init` or `init_on_batch` first.
* `predict` now has an optional `initialize` argument if you do want to force initialization of a network-only Model (no metrics or optimizer).

A typical example would look like this:

```python
model = elegy.Model(...)

# call init at the beggining
model.init(X_train, y_train)

# now you can call predict
y_pred = model.predict(X_train[:64])

# call fit as usual
history = model.fit(
    x=X_train,
    y=y_train,
    ....
)
```

I hope this change makes it easier to deal with complex states (e.g. distributed states for `pmap`) by reducing the complexity around initialization at the price of a maximum of just 1 additional line of code from the user.

### CODE REVIEW

**Comment by cgarciae** [comment]
@charlielito @alexander-g

**Comment by alexander-g** [comment]
> `model.init(X_train, y_train)`

Is `y_train` required? What if I don't have it or don't know what it should look like, maybe because I want to use a module written by another author? Consider an object detector like FasterRCNN which has quite some output heads (8x) of different shapes, the user would need to find out the output structure then construct a conforming `y_train` before they can call `init`.

As you may have noticed before, I'm all in to reduce boilerplate and make the user's life easier, this PR complicates it. The largest benefit I see here is the reduction of initialization/compilation time, but isn't `init` called without `jit` anyway?

**Comment by cgarciae** [comment]
> Is y_train required?
It is only required if you have metrics that use `y_true`.

> the user would need to find out the output structure then construct a conforming y_train before they can call init. 

Since you are just exploring the output structure and won't actually train the model, for this specific use case you can use:

```
y_pred = model.predict(X_train, initialize=True)
```
This flag was added for when you are doing parameter transfer and explicitly opt-in to only initialize the network. This will crash if you have a metric.

We can also add a `Model.eval_shape` method that calls `pred_step` with `jax.eval_shape` and doesn't require initialization.

**Comment by alexander-g** [comment]
Ok, sounds good. Maybe I misunderstood.
